### PR TITLE
[C++20] [Modules] Introduce -fexperimental-modules-reduced-bmi

### PR DESCRIPTION
**Pull request by ChuanqiXu9**
This is the driver part of https://github.com/llvm/llvm-project/pull/75894.

This patch introduces '-fexperimental-modules-reduced-bmi' to enable generating the reduced BMI.

This patch did:
- When `-fexperimental-modules-reduced-bmi` is specified but `--precompile` is not specified for a module unit, we'll skip the precompile phase to avoid unnecessary  two-phase compilation phases. Then if `-c` is specified, we will generate the reduced BMI in CodeGenAction as a by-product.
- When `-fexperimental-modules-reduced-bmi` is specified and `--precompile` is specified, we will generate the reduced BMI in GenerateModuleInterfaceAction as a by-product.
- When `-fexperimental-modules-reduced-bmi` is specified for a non-module unit. We don't do anything nor try to give a warn. This is more user friendly so that the end users can try to test and experiment with the feature without asking help from the build systems.

The core design idea is that users should be able to enable this easily with the existing cmake mechanisms.

The future plan for the flag is:
- Add this to clang19 and make it opt-in for 1~2 releases. It depends on the testing feedback to decide how long we like to make it opt-in.
- Then we can announce the existing BMI generating may be deprecated and suggesting people (end users or build systems) to enable this for 1~2 releases.
- Finally we will enable this by default. When that time comes, the term `BMI` will refer to the reduced BMI today and the existing BMI will only be meaningful to build systems which loves to support two phase compilations.

I'll send release notes and document  in seperate commits after this get landed.

### CODE REVIEW

**Comment by koplas** [comment]
Using `-fgen-reduced-bmi` currently fails for me with a case that looks like this:
```C++
// a.hpp
  template <typename T>
  void ignore(T const &) noexcept {}
  
  inline void resultCheck(char const *message) {
      ignore(message);
  }
  
// b.cppm
    module;
    #include "a.hpp"
    export module b;
    
    export {
        using ignore;
        using resultCheck;
    }
    
// c.cppm
    export module c;
    import b;
    
    export void test() {
        resultCheck(nullptr);
    }
```
This will result in this linker error: `undefined reference to void ignore<char const*>(char const* const&)`.
Removing `-fgen-reduced-bmi` will result in linking without errors. Do I need to pass certain flags to CMake?

**Comment by koplas** [comment]
Do I miss something? The performance and file size is similar with and without `-fgen-reduced-bmi`.
To reproduce clone https://github.com/koplas/clang-modules-test and run `build.sh` and `build_thin_bmi.sh`.

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -3031,6 +3032,11 @@ defm skip_odr_check_in_gmf : BoolOption<"f", "skip-odr-check-in-gmf",
           "Perform ODR checks for decls in the global module fragment.">>,
   Group<f_Group>;
 
+def gen_reduced_bmi : Flag<["-"], "fgen-reduced-bmi">,
+  Group<i_Group>, Visibility<[ClangOption, CC1Option]>,
```
Comment (MaskRay) [inline clang/include/clang/Driver/Options.td:3036]: should be f_Group

File: clang/test/Driver/module-fgen-reduced-bmi.cppm [hunk 2]
```diff
@@ -0,0 +1,53 @@
+// It is annoying to handle different slash direction
+// in Windows and Linux. So we disable the test on Windows
+// here.
+// REQUIRES: !system-windows
+// On AIX, the default output for `-c` may be `.s` instead of `.o`,
+// which makes the test fail. So disable the test on AIX.
+// REQUIRES: !system-aix
+//
+// RUN: rm -rf %t
```
Comment (MaskRay) [inline clang/test/Driver/module-fgen-reduced-bmi.cppm:9]: redundant `mkdir`

`// RUN: rm -rf %t && split-file %s %t && cd %t`

If Windows backslashes feel annoying, you can add a `cd %t`

File: clang/test/Driver/module-fgen-reduced-bmi.cppm [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// It is annoying to handle different slash direction
+// in Windows and Linux. So we disable the test on Windows
+// here.
+// REQUIRES: !system-windows
```
Comment (MaskRay) [inline clang/test/Driver/module-fgen-reduced-bmi.cppm:4]: UNSUPPORTED is much more common

File: clang/include/clang/Driver/Options.td [hunk 4]
```diff
@@ -3031,6 +3032,11 @@ defm skip_odr_check_in_gmf : BoolOption<"f", "skip-odr-check-in-gmf",
           "Perform ODR checks for decls in the global module fragment.">>,
   Group<f_Group>;
 
+def gen_reduced_bmi : Flag<["-"], "fgen-reduced-bmi">,
```
Comment (iains) [inline clang/include/clang/Driver/Options.td:3035]: If this is going to be a user-facing flag (even for a few releases) then I think it should be spelled in a way that make it modules-relatated (e.g. -fmodules-reduce-bmi, or -fmodules-minimise-bmi),  as you know I'm not a fan of increasing the already huge number of modules-related user-facing options... (but I'll comment on that separately)

**Review by iains — COMMENTED** [review]
I have no issue with the general intention or phasing here; my main concern is that we are introducing [yet] another user-facing modules option that actually we intend will become irrelevant after a few releases.

In practice, removing user-facing options is not very easy.

So:
  1. Perhaps we could have only the cc1 option and you could introduce some temporary handling in the driver to recognise it and amend the --precompile behaviour?
  2. Maybe the project could introduce something like -fexperimental-xxxx-xxxx with a clear statement that fexperimental flags cannot be relied on to be stable (or even present) in any future release)?

**Comment by dwblaikie** [comment]
+1 to @iains's comments about being careful about the introduction and naming of driver flags & probably avoid it in this case, if possible, or try to make it clearly experimental.

File: clang/test/Driver/module-fgen-reduced-bmi.cppm (hunk 3 above)
Comment (ChuanqiXu9) [reply clang/test/Driver/module-fgen-reduced-bmi.cppm:4]: Done

File: clang/include/clang/Driver/Options.td (hunk 4 above)
Comment (ChuanqiXu9) [reply clang/include/clang/Driver/Options.td]: Done by using `-fmodules-reduce-bmi`

**Comment by ChuanqiXu9** [comment]
@iains @dwblaikie Understood. And I thought the major problem is that there are a lot flags from clang modules. And it is indeed confusing. But given we have to introduce new flags in this case, probably we can only try to make it more clear by better documents.

**Comment by ChuanqiXu9** [comment]
> Using `-fgen-reduced-bmi` currently fails for me with a case that looks like this:
> 
> ```c++
> // a.hpp
>   template <typename T>
>   void ignore(T const &) noexcept {}
>   
>   inline void resultCheck(char const *message) {
>       ignore(message);
>   }
>   
> // b.cppm
>     module;
>     #include "a.hpp"
>     export module b;
>     
>     export {
>         using ignore;
>         using resultCheck;
>     }
>     
> // c.cppm
>     export module c;
>     import b;
>     
>     export void test() {
>         resultCheck(nullptr);
>     }
> ```
> 
> This will result in this linker error: `undefined reference to void ignore<char const*>(char const* const&)`. Removing `-fgen-reduced-bmi` will result in linking without errors. Do I need to pass certain flags to CMake?

This may be fixed in trunk. This should be a bug in Reduced BMI. Note that the current patch is only the driver part of Reduced BMI. And this is the reason why we need a new flag. Otherwise I'll propose to enable it by default : )

**Comment by ChuanqiXu9** [comment]
> Do I miss something? The performance and file size is similar with and without `-fgen-reduced-bmi`. To reproduce clone https://github.com/koplas/clang-modules-test and run `build.sh` and `build_thin_bmi.sh`.

Currently, the Reduced BMI will only remove function bodies for non-inline functions and definitions for some variables...

And declarations in headers or GMF are primarily inline, so probably the current Reduced BMI won't produce a strong impact on modules whose contents are primarily in GMF as expectedly.

For example, I can observe some impact on https://github.com/alibaba/async_simple/tree/CXX20Modules, where the repo has more use codes in the module purview instead of the GMF.

Further, in my mind, we can reduce more things in Reduced BMI. An example may be https://github.com/llvm/llvm-project/pull/76930.

Thanks for testing : )

**Comment by iains** [comment]
> @iains @dwblaikie Understood. And I thought the major problem is that there are a lot flags from clang modules. And it is indeed confusing. But given we have to introduce new flags in this case, probably we can only try to make it more clear by better documents.

So you do not think it possible to restrict the new flag to be "internal" (i.e. cc1-only) and to put some _temporary_ driver processing to handle that? (I agree that this is an unusual mechanism, but the idea is to recognise that the driver-side processing is only expected to me temporary).

**Comment by ChuanqiXu9** [comment]
> > @iains @dwblaikie Understood. And I thought the major problem is that there are a lot flags from clang modules. And it is indeed confusing. But given we have to introduce new flags in this case, probably we can only try to make it more clear by better documents.
> 
> So you do not think it possible to restrict the new flag to be "internal" (i.e. cc1-only) and to put some _temporary_ driver processing to handle that? (I agree that this is an unusual mechanism, but the idea is to recognise that the driver-side processing is only expected to me temporary).

I have no idea how can we make that. We still need the users to provide something to enable reduced BMI. And I think it is symmetric to a new flag.

**Comment by iains** [comment]
> > > @iains @dwblaikie Understood. And I thought the major problem is that there are a lot flags from clang modules. And it is indeed confusing. But given we have to introduce new flags in this case, probably we can only try to make it more clear by better documents.
> > 
> > 
> > So you do not think it possible to restrict the new flag to be "internal" (i.e. cc1-only) and to put some _temporary_ driver processing to handle that? (I agree that this is an unusual mechanism, but the idea is to recognise that the driver-side processing is only expected to me temporary).
> 
> I have no idea how can we make that. We still need the users to provide something to enable reduced BMI. And I think it is symmetric to a new flag.

What I mean is that (a) we need the internal 'cc1' flag permanently; but (b) we do not expect to need a user-facing driver flag permanently.  and (c) We want to allow users to try this out.

I am suggesting we could say "to try this out use -Xclang -fmodules-reduced-bmi" and have _temporary_ code in the driver to deal with the changes needed to phasing.

If this is not possible. then I suppose I am a bit sad that we keep saying 'there are too many modules options' - but yet still add more.   however - we need to make progress, so if the suggestion here is really a non-starter .. then such is life.

Perhaps the second suggestion (-fexperimental-xxxxx options) could be discussed at the project level.

**Comment by ChuanqiXu9** [comment]
> > > > @iains @dwblaikie Understood. And I thought the major problem is that there are a lot flags from clang modules. And it is indeed confusing. But given we have to introduce new flags in this case, probably we can only try to make it more clear by better documents.
> > > 
> > > 
> > > So you do not think it possible to restrict the new flag to be "internal" (i.e. cc1-only) and to put some _temporary_ driver processing to handle that? (I agree that this is an unusual mechanism, but the idea is to recognise that the driver-side processing is only expected to me temporary).
> > 
> > 
> > I have no idea how can we make that. We still need the users to provide something to enable reduced BMI. And I think it is symmetric to a new flag.
> 
> What I mean is that (a) we need the internal 'cc1' flag permanently; but (b) we do not expect to need a user-facing driver flag permanently. and (c) We want to allow users to try this out.
> 
> I am suggesting we could say "to try this out use -Xclang -fmodules-reduced-bmi" and have _temporary_ code in the driver to deal with the changes needed to phasing.
> 
> If this is not possible. then I suppose I am a bit sad that we keep saying 'there are too many modules options' - but yet still add more. however - we need to make progress, so if the suggestion here is really a non-starter .. then such is life.
> 
> Perhaps the second suggestion (-fexperimental-xxxxx options) could be discussed at the project level.

Got your point. But I feel the `-Xclang xxx` style or the `-fexperimental-xxx` style may make people feel, "oh, this is not ready, let's wait". Then we may lose more testing opportunity. Also I feel such options are symmetric to the existing one. It still makes the interfaces to be more complicated. I don't feel they are better than the existing one.

And yes, it is indeed that we have a lot modules options. But I think this may not be the blocking point for adding new modules options if we think it is really needed. If there are things needed to be added by flags really and we don't do that due to we feel we have too many flags, I feel it is somewhat "not eating for fear of choking". I think what we need to do for this case is to be pretty careful when adding new flags and I think we already make it for Reduced BMI.

**Comment by dwblaikie** [comment]
> > > > > @iains @dwblaikie Understood. And I thought the major problem is that there are a lot flags from clang modules. And it is indeed confusing. But given we have to introduce new flags in this case, probably we can only try to make it more clear by better documents.
> > > > 
> > > > 
> > > > So you do not think it possible to restrict the new flag to be "internal" (i.e. cc1-only) and to put some _temporary_ driver processing to handle that? (I agree that this is an unusual mechanism, but the idea is to recognise that the driver-side processing is only expected to me temporary).
> > > 
> > > 
> > > I have no idea how can we make that. We still need the users to provide something to enable reduced BMI. And I think it is symmetric to a new flag.
> > 
> > 
> > What I mean is that (a) we need the internal 'cc1' flag permanently; but (b) we do not expect to need a user-facing driver flag permanently. and (c) We want to allow users to try this out.
> > I am suggesting we could say "to try this out use -Xclang -fmodules-reduced-bmi" and have _temporary_ code in the driver to deal with the changes needed to phasing.
> > If this is not possible. then I suppose I am a bit sad that we keep saying 'there are too many modules options' - but yet still add more. however - we need to make progress, so if the suggestion here is really a non-starter .. then such is life.
> > Perhaps the second suggestion (-fexperimental-xxxxx options) could be discussed at the project level.
> 
> Got your point. But I feel the `-Xclang xxx` style or the `-fexperimental-xxx` style may make people feel, "oh, this is not ready, let's wait". Then we may lose more testing opportunity. 

That seems good to me - these are pretty experimental directions we're going in. We haven't figured out how this stuff should work long term - we are experimenting.

> Also I feel such options are symmetric to the existing one.

What do you mean by "symmetric"?

The difference @iains is trying to highlight is that adding driver flags is a long-term commitment burden - adding cc1 flags, clarifying that these are not long term supported/guaranteed parts of the interface, but a way to experiment with some things until we figure out what the long term supported interface is.

Especially if we think the long-term future is always (or, at least by default) reduced BMIs, experimenting with it under a flag until we have confidence in that direction - then maybe just changing the Clang behavior (again, perhaps with a cc1 flag to opt-out as an escape hatch that's intended to be short-term while we figure out whatever bugs lead to the need to use that escape hatch - then when we fixt he bugs and remove the flag, the people on the bleeding edge will need to remove the flag, which is fine/good - rather than leaving these weird flags around forever).

**Comment by ChuanqiXu9** [comment]
Got it. I've renamed the flag as `-fexperimental-modules-reduced-bmi`. I feel the suggestion like let users to use `-Xclang` options look odd..

---

> What do you mean by "symmetric"?

I mean, we always want the users to do something explicitly to enable the feature. But this might not be important now.

**Comment by iains** [comment]
> Got it. I've renamed the flag as `-fexperimental-modules-reduced-bmi`.

Thanks.

> I feel the suggestion like let users to use `-Xclang` options look odd..

I think the point here is that we want **expert** users to try this out (with understanding that it might not behave exactly as they expect).  Because it is experimental, it is not yet ready for "end users" - I would expect expert users to be able to deal with -Xclang xxxxx  (but I will not insist, it's just a suggestion)

**Comment by ChuanqiXu9** [comment]
> > Got it. I've renamed the flag as `-fexperimental-modules-reduced-bmi`.
> 
> Thanks.
> 
> > I feel the suggestion like let users to use `-Xclang` options look odd..
> 
> I think the point here is that we want **expert** users to try this out (with understanding that it might not behave exactly as they expect). Because it is experimental, it is not yet ready for "end users" - I would expect expert users to be able to deal with -Xclang xxxxx (but I will not insist, it's just a suggestion)

Agreed.

**Comment by ChuanqiXu9** [comment]
I'd like to land this patch in next week if no objection comes in.

File: clang/lib/Driver/ToolChains/Clang.h [hunk 5]
```diff
@@ -193,6 +193,20 @@ DwarfFissionKind getDebugFissionKind(const Driver &D,
...
+// - If `-fmodule-output=` is specfied, then the module file is
+//   writing to the value.
+// - Otherwise if the output object file of the module unit is specified, the
+// output path
+//   of the module file should be the same with the output object file except
+//   the corresponding suffix. This requires both `-o` and `-c` are specified.
+// - Otherwise, the output path of the module file will be the same with the
+//   input with the corresponding suffix.
+llvm::SmallString<256>
+getCXX20NamedModuleOutputPath(const llvm::opt::ArgList &Args,
```
Comment (dwblaikie) [inline clang/lib/Driver/ToolChains/Clang.h:208]: Pulling this work out into a named function could probably be a preliminary NFC commit?

File: clang/include/clang/Driver/Options.td [hunk 6]
```diff
@@ -3017,6 +3017,7 @@ defm prebuilt_implicit_modules : BoolFOption<"prebuilt-implicit-modules",
 
 def fmodule_output_EQ : Joined<["-"], "fmodule-output=">,
   Flags<[NoXarchOption]>, Visibility<[ClangOption, CC1Option]>,
+  MarshallingInfoString<FrontendOpts<"ModuleOutputPath">>,
```
Comment (dwblaikie) [inline clang/include/clang/Driver/Options.td:3020]: What's the motivation to include this change in this patch? Could it be separated?

File: clang/lib/Driver/ToolChains/Clang.h (hunk 5 above)
Comment (ChuanqiXu9) [reply clang/lib/Driver/ToolChains/Clang.h]: Done in https://github.com/llvm/llvm-project/commit/21f85e230056172cffcaec76352e5a2019b54b86

File: clang/include/clang/Driver/Options.td (hunk 6 above)
Comment (ChuanqiXu9) [reply clang/include/clang/Driver/Options.td:3020]: Since the flag was only used in drivers. So we didn't need a variable. But after this patch, we need a variable. If we separate this out, we may get a unused variable in that patch. I feel it is not so good.

File: clang/lib/Serialization/GeneratePCH.cpp [hunk 7]
```diff
@@ -101,12 +102,26 @@ ReducedBMIGenerator::ReducedBMIGenerator(const Preprocessor &PP,
...
   Module *M = Ctx.getCurrentNamedModule();
-  assert(M->isNamedModuleUnit() &&
+  assert(M && M->isNamedModuleUnit() &&
          "ReducedBMIGenerator should only be used with C++20 Named modules.");
   return M;
 }
 
 void ReducedBMIGenerator::HandleTranslationUnit(ASTContext &Ctx) {
+  // FIMXE: We'd better to wrap such options to a new class ASTWriterOptions.
+  getPreprocessor()
```
Comment (dwblaikie) [inline clang/lib/Serialization/GeneratePCH.cpp:112]: What's the motivation for this change/how does it relate to the purpose of this patch?

(& it might be easier to read if the HeaderSearchOpts were accessed once, stored in a reference, then the various settings were made:
```
HeaderSearchOptions& Opts = getPreprocessor().getHeaderSearchInfo().getHeaderSearchOpts();
Opts.ModulesSkipDiagnosticOptions = true;
Opts.ModulesSkipHeaderSearchPaths = true;
Opts.ModulesSkipPragmaDiagnosticMappings = true;
```
)

Comment (ChuanqiXu9) [reply clang/lib/Serialization/GeneratePCH.cpp]: Oh, my bad. The motivation is that during my local testing, I found the size of the reduced BMI may be larger than the full BMI, which should be incorrect. And then I tried to fix this on the fly... I've splitted this into a NFC patch https://github.com/llvm/llvm-project/commit/ed1cfffe9b2b2d3cc9279ff83400ace156b317a2

**Comment by ChuanqiXu9** [comment]
Rebased with main. @dwblaikie may you have more comments?

File: clang/lib/CodeGen/CodeGenAction.cpp [hunk 8]
```diff
@@ -1061,6 +1070,16 @@ CodeGenAction::CreateASTConsumer(CompilerInstance &CI, StringRef InFile) {
...
 
+  if (CI.getFrontendOpts().GenReducedBMI &&
+      !CI.getFrontendOpts().ModuleOutputPath.empty()) {
+    std::vector<std::unique_ptr<ASTConsumer>> Consumers{2};
+    Consumers[0] = std::make_unique<ReducedBMIGenerator>(
+        CI.getPreprocessor(), CI.getModuleCache(),
+        CI.getFrontendOpts().ModuleOutputPath);
+    Consumers[1] = std::move(Result);
+    return std::make_unique<MultiplexConsumer>(std::move(Consumers));
+  }
```
Comment (dwblaikie) [inline clang/lib/CodeGen/CodeGenAction.cpp:1081]: Why is this new code, rather than an alternative codepath of existing code? (like some code that was producing the BMI would go from producing the unreduced BMI to producing the reduced BMI)

Figuring this out would help me unedrstand why `ModuleOutputPath` is new (this patch doesn't add the functionality to output a module to a path - we had that already, this just changes the content that goes there)

I guess because maybe this patch is changing the way we output the BMI - or that the reduced BMI is being output in a different way than the non-reduced BMI?

Should we address that? Make the current full BMI emission work the same way as the reduced BMI? (or, more precisely, in a pre-patch, change the full BMI emission to use a mechanism that can then be reused for the reduced BMI?)

**Review by dwblaikie — APPROVED** [review]
Give it a go

File: clang/lib/CodeGen/CodeGenAction.cpp [hunk 9]
```diff
@@ -1061,6 +1070,16 @@ CodeGenAction::CreateASTConsumer(CompilerInstance &CI, StringRef InFile) {
     CI.getPreprocessor().addPPCallbacks(std::move(Callbacks));
   }
 
+  if (CI.getFrontendOpts().GenReducedBMI &&
+      !CI.getFrontendOpts().ModuleOutputPath.empty()) {
+    std::vector<std::unique_ptr<ASTConsumer>> Consumers{2};
```
Comment (mizvekov) [inline clang/lib/CodeGen/CodeGenAction.cpp:1075]: Minor nit: https://llvm.org/docs/CodingStandards.html#id28
```suggestion
    std::vector<std::unique_ptr<ASTConsumer>> Consumers(2);
```

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 10]
```diff
@@ -4045,6 +4045,24 @@ static bool RenderModulesOptions(Compilation &C, const Driver &D,
...
+      (Input.getType() == driver::types::TY_CXXModule ||
+       Input.getType() == driver::types::TY_PP_CXXModule)) {
+    CmdArgs.push_back("-fexperimental-modules-reduced-bmi");
+
+    if (Args.hasArg(options::OPT_fmodule_output_EQ))
+      Args.AddLastArg(CmdArgs, options::OPT_fmodule_output_EQ);
+    else
+      CmdArgs.push_back(Args.MakeArgString(
+          "-fmodule-output=" +
+          getCXX20NamedModuleOutputPath(Args, Input.getBaseInput())));
```
Comment (mizvekov) [inline clang/lib/Driver/ToolChains/Clang.cpp:4058]: Would it be reasonable to push two separate arguments, instead of concatenating them?

**Review by mizvekov — APPROVED** [review]
I share the objections that it may be too soon to introduce a driver flag for this. Only a frontend flag is ok for now, but it's non blocking on my side because it doesn't look like it will be particularly hard to deprecate it later.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 10 above)
Comment (ChuanqiXu9) [reply clang/lib/Driver/ToolChains/Clang.cpp:4058]: I am not sure. I tried to not concat it but the test gets failing. I didn't see why but I see almost we always concat it in such cases (appending a variable to some flag ends with `=`). So I feel it may not be problematic or we can fix such things together if we want.

File: clang/lib/CodeGen/CodeGenAction.cpp (hunk 8 above)
Comment (ChuanqiXu9) [reply clang/lib/CodeGen/CodeGenAction.cpp]: For full BMI, the workflow is:

```
.cpp -> .pcm -> .o
```

no matter if `-fmodule-output` is specified or not. If `-fmodule-output` is not specified, the `.pcm` file will be generated in `/tmp` directory.

And for reduced BMI, the workflow is:

```
.cpp ---------> .o
               -----> .pcm
```

that said, we generate the .o directly from the .cpp then we don't need the .pcm to contain the full information to generate the .o . 

Then for the original question, can we make the full BMI in the same way with the reduced BMI? Maybe we can, but it looks like an overkill. Since if the .o don't need the .pcm, we don't need the .pcm to be full.

**Comment by ChuanqiXu9** [comment]
Thanks for reviewing.
